lib/public_key/src/pubkey_policy_tree.erl

```diff
@@ -0,0 +1,313 @@
+    end.
+
+%%--------------------------------------------------------------------
+%-spec prune_invalid_nodes(policy_tree(), [policy_node()]) -> policy_tree().
+%%
+%% Delete all invalid policy nodes and thier children.
+%%--------------------------------------------------------------------
+prune_invalid_nodes(Tree, []) ->
+    Tree;
+prune_invalid_nodes({Root, Children}, InvalidNodes) ->
+    case prune_invalid_nodes(Children, InvalidNodes) of
+        [] ->
+         empty();
+        NewChildren ->
+            {Root, NewChildren}
+    end;
+prune_invalid_nodes(TreeBottom, InvalidNodes) when is_list(TreeBottom)->
+    lists:filtermap(fun({Parent, Children}) ->
+                             case keep_policy_node(Parent, InvalidNodes) of
+                                 false ->
+                                     false;
+                                 {true, Parent} ->
+                                     {true, {Parent, prune_invalid_nodes(Children, InvalidNodes)}}
+                             end;
+                       (#{} = Child) ->
+                            keep_policy_node(Child, InvalidNodes)
+                    end, TreeBottom).
+    
+%%--------------------------------------------------------------------
+-spec policy_node(public_key:oid(), term(), [public_key:oid()]) -> policy_node().
+%%
+%% Creates a policy node
+%%--------------------------------------------------------------------
+policy_node(ValidPolicy, Qualifiers, ExpPolicySet) ->
+    QualifierSet = format_qualifers(Qualifiers),
+    #{valid_policy => ValidPolicy,
+      qualifier_set => QualifierSet,
+      expected_policy_set => ExpPolicySet}.
+
+%%--------------------------------------------------------------------
+-spec root() -> policy_tree().
+%%
+%% Create initial root node used as start value when building the policy tree.
+%%--------------------------------------------------------------------
+root() ->
+    {any_node(), []}.
+
+%%--------------------------------------------------------------------
+-spec valid_policy_node_set(policy_tree()) -> [policy_node()].
+%%
+%% Determine the set of policy nodes whose parent nodes have a
+%% valid_policy of anyPolicy.  This is the valid_policy_node_set.
+%%--------------------------------------------------------------------
+valid_policy_node_set({#{valid_policy := ?anyPolicy}, [{_, _}| _]= Children0}) ->
+    Parents = [Parent || {Parent, _} <- Children0],
+    Children = [Child || {_ , Child} <- Children0],
+    Parents ++ valid_policy_node_set(Parents, Children);
+valid_policy_node_set({#{valid_policy := ?anyPolicy}, Children}) ->
+     Children.
+valid_policy_node_set([#{valid_policy := ?anyPolicy} = Parent | PRest], [Children | CRest]) ->
+    valid_policy_node_set({Parent, Children}) ++ valid_policy_node_set(PRest, CRest);
+valid_policy_node_set([], []) ->
+    [];
+valid_policy_node_set([_Parent | PRest], [_Children | CRest]) ->
+    valid_policy_node_set(PRest, CRest).
+
```

Comment:
Simplify to something like this? 
```
valid_policy_node_set(Tree) ->
     valid_policy_node_set(Tree, Acc).

valid_policy_node_set({#{valid_policy := ?anyPolicy}, [{_, _}| _]= Children0}, Acc) ->
    Parents = [Parent || {Parent, _} <- Children0],
    lists:foldl(fun valid_policy_node_set/2, Parents ++ Acc, Children0);
valid_policy_node_set({#{valid_policy := ?anyPolicy}, Children}, Acc) when is_list(Children0) ->
    Children ++ Acc;
valid_policy_node_set(_) ->
    [].
```

Reply:
I will investigate!

Reply:
@dgud This just is not equivalent. I added some comments instead. What do you think?

Reply:
New suggestion included :)